### PR TITLE
 *: add split index region syntax

### PR DESCRIPTION
**Pull request by crazycs520**
### What problem does this PR solve? 
PR: https://github.com/pingcap/tidb/pull/10138 support pre-split table region. But it will only split 1 region for per index. Then the index region may still be the hot region and the write bottleneck.

### What is changed and how it works?
This PR support split index region by SQL.  Syntax like below: 

```sql
SPLIT TABLE table_name INDEX index_name BY (val1...), (val2...),(val3...) ...
```

**example**
```sql
split table t1 idx1 (100), (1000), (10000);   
//range is : 0~100, 100 ~1000, 1000~ 10000 ， 10000 ~ maxInt64

split table t1 idx2 ("a", "2000-01-01 00:00:01"),  ("b", "2019-04-17 14:26:19"),  ("c", "");  
// range is: minIndexValue ~ ("a", "2000-01-01 00:00:01"),
("a", "2000-01-01 00:00:01") ~ ("b", "2019-04-17 14:26:19"),
("b", "2019-04-17 14:26:19") ~ ("c", ""),
("c", "") ~ maxIndexValue.
```

Please merge  https://github.com/pingcap/tidb/pull/10138 firstly.
Related parser PR: https://github.com/pingcap/parser/pull/297

### Check List 

Tests 

 - Adding test.

Code changes

 - Has exported function/method change

Side effects

Related changes

### CODE REVIEW

File: executor/split.go [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+func (e *SplitIndexRegionExec) Next(ctx context.Context, _ *chunk.RecordBatch) error {
+	store := e.ctx.GetStore()
+	s, ok := store.(splitableStore)
+	if !ok {
+		return nil
+	}
+	regionIDs := make([]uint64, 0, len(e.valueLists))
+	index := tables.NewIndex(e.table.Meta().ID, e.table.Meta(), e.indexInfo)
+	for _, values := range e.valueLists {
+		idxKey, _, err := index.GenIndexKey(e.ctx.GetSessionVars().StmtCtx, values, 1, nil)
```
Comment (nolouch) [inline executor/split.go]: `1` means min handle id?

File: store/tikv/split_region.go [hunk 2]
```diff
@@ -16,7 +16,6 @@ package tikv
 import (
 	"bytes"
 	"context"
-
```
Comment (zimulala) [inline store/tikv/split_region.go]: Is it a misoperation？

File: planner/core/planbuilder.go [hunk 3]
```diff
@@ -1588,6 +1590,54 @@ func (b *PlanBuilder) buildLoadStats(ld *ast.LoadStatsStmt) Plan {
...
+	indexInfo := tblInfo.FindIndexByName(strings.ToLower(node.IndexName))
+	if indexInfo == nil {
+		return nil, ErrKeyDoesNotExist.GenWithStackByArgs(node.IndexName, tblInfo.Name)
+	}
+	schema := expression.TableInfo2SchemaWithDBName(b.ctx, node.Table.Schema, tblInfo)
+	mockTablePlan := LogicalTableDual{}.Init(b.ctx)
+	mockTablePlan.SetSchema(schema)
+
+	tps := make([]*types.FieldType, 0, len(indexInfo.Columns))
+	for _, c := range indexInfo.Columns {
```
Comment (winkyao) [inline planner/core/planbuilder.go]: why we need this tps array?

File: planner/core/planbuilder.go [hunk 4]
```diff
@@ -1588,6 +1590,54 @@ func (b *PlanBuilder) buildLoadStats(ld *ast.LoadStatsStmt) Plan {
...
+		if len(valuesItem) > len(indexInfo.Columns) {
+			return nil, ErrWrongValueCountOnRow.GenWithStackByArgs(i + 1)
+		}
+		valueList := make([]types.Datum, 0, len(valuesItem))
+		for j, valueItem := range valuesItem {
+			x, ok := valueItem.(*driver.ValueExpr)
+			if !ok {
+				return nil, errors.Trace(errors.New("expect constant values"))
+			}
+			value, err := x.Datum.ConvertTo(b.ctx.GetSessionVars().StmtCtx, tps[j])
```
Comment (winkyao) [inline planner/core/planbuilder.go]: tps[j] can be `tblInfo.Columns[j].FieldType`

File: store/tikv/split_region.go (hunk 2 above)
Comment (crazycs520) [reply store/tikv/split_region.go]: ye...

File: planner/core/planbuilder.go (hunk 4 above)
Comment (crazycs520) [reply planner/core/planbuilder.go]: `tblInfo.Columns[indexInfo.Columns[j].Offset]`
done.

File: executor/split.go (hunk 1 above)
Comment (crazycs520) [reply executor/split.go]: yes, Maybe `0` is better?

Comment (winkyao) [reply executor/split.go]: you can use a constant, better than hard code!

Comment (zimulala) [reply executor/split.go]: @winkyao
I think `1` isn't always the min handle. If a table has a PKIsHandle column, then the handle may be a negative value.
So I think we need to use `math.MinInt64`.

File: executor/executor.go [hunk 5]
```diff
@@ -1379,6 +1379,10 @@ func ResetContextOfStmt(ctx sessionctx.Context, s ast.StmtNode) (err error) {
 			sc.InShowWarning = true
 			sc.SetWarnings(vars.StmtCtx.GetWarnings())
 		}
+	case *ast.SplitIndexRegionStmt:
+		sc.IgnoreTruncate = false
+		sc.IgnoreZeroInDate = true
+		sc.AllowInvalidDate = vars.SQLMode.HasAllowInvalidDatesMode()
```
Comment (winkyao) [inline executor/executor.go]: What is this field used for?

File: executor/split.go (hunk 1 above)
Comment (winkyao) [reply executor/split.go]: @zimulala The key is encoding in binary format, so negative value is always greater than a positive value. @nolouch Right?

File: executor/executor.go (hunk 5 above)
Comment (crazycs520) [reply executor/executor.go]: Field `sc.AllowInvalidDate` just keep consistent with the default, I just want to set `IgnoreTruncate` to false here.

File: planner/core/planbuilder.go [hunk 6]
```diff
@@ -1588,6 +1590,48 @@ func (b *PlanBuilder) buildLoadStats(ld *ast.LoadStatsStmt) Plan {
...
+	indexValues := make([][]types.Datum, 0, len(node.ValueLists))
+	for i, valuesItem := range node.ValueLists {
+		if len(valuesItem) > len(indexInfo.Columns) {
+			return nil, ErrWrongValueCountOnRow.GenWithStackByArgs(i + 1)
+		}
+		valueList := make([]types.Datum, 0, len(valuesItem))
+		for j, valueItem := range valuesItem {
+			x, ok := valueItem.(*driver.ValueExpr)
+			if !ok {
+				return nil, errors.Trace(errors.New("expect constant values"))
```
[the remark this answers is not in the document]
Comment (crazycs520) [reply planner/core/planbuilder.go]: done.

File: executor/split.go (hunk 1 above)
Comment (crazycs520) [reply executor/split.go]: Here is a test:
0                        encode to byte is: [128 0 0 0 0 0 0 0]
math.MinInt64 encode to byte is:  [128 0 0 0 0 0 0 0]

so,  math.MinInt64 is better.

File: planner/core/planbuilder.go (hunk 3 above)
Comment (crazycs520) [reply planner/core/planbuilder.go]: done.

File: planner/core/planbuilder.go [hunk 7]
```diff
@@ -1605,6 +1607,48 @@ func (b *PlanBuilder) buildLoadStats(ld *ast.LoadStatsStmt) Plan {
...
+		if len(valuesItem) > len(indexInfo.Columns) {
+			return nil, ErrWrongValueCountOnRow.GenWithStackByArgs(i + 1)
+		}
+		valueList := make([]types.Datum, 0, len(valuesItem))
+		for j, valueItem := range valuesItem {
+			x, ok := valueItem.(*driver.ValueExpr)
+			if !ok {
+				return nil, errors.New("expect constant values")
+			}
+			value, err := x.Datum.ConvertTo(b.ctx.GetSessionVars().StmtCtx, &tblInfo.Columns[indexInfo.Columns[j].Offset].FieldType)
```
Comment (winkyao) [inline planner/core/planbuilder.go]: colOffset := indexInfo.Columns[j].Offset
value, err := x.Datum.ConvertTo(b.ctx.GetSessionVars().StmtCtx, &tblInfo.Columns[colOffset].FieldType)

File: planner/core/planbuilder.go [hunk 8]
```diff
@@ -1605,6 +1607,48 @@ func (b *PlanBuilder) buildLoadStats(ld *ast.LoadStatsStmt) Plan {
...
 }
 
+func (b *PlanBuilder) buildSplitIndexRegion(node *ast.SplitIndexRegionStmt) (Plan, error) {
+	tblInfo := node.Table.TableInfo
+	indexInfo := tblInfo.FindIndexByName(strings.ToLower(node.IndexName))
+	if indexInfo == nil {
+		return nil, ErrKeyDoesNotExist.GenWithStackByArgs(node.IndexName, tblInfo.Name)
+	}
+	schema := expression.TableInfo2SchemaWithDBName(b.ctx, node.Table.Schema, tblInfo)
+	mockTablePlan := LogicalTableDual{}.Init(b.ctx)
```
Comment (winkyao) [inline planner/core/planbuilder.go]: What is mockTablePlan used for?

Comment (crazycs520) [reply planner/core/planbuilder.go]: It is useless now. Already delete this.

File: planner/core/planbuilder.go (hunk 7 above)
Comment (crazycs520) [reply planner/core/planbuilder.go]: done

File: executor/split.go [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+type splitableStore interface {
+	SplitRegionAndScatter(splitKey kv.Key) (uint64, error)
+	WaitScatterRegionFinish(regionID uint64) error
+}
+
+// Next implements the Executor Next interface.
+func (e *SplitIndexRegionExec) Next(ctx context.Context, _ *chunk.RecordBatch) error {
+	store := e.ctx.GetStore()
+	s, ok := store.(splitableStore)
+	if !ok {
```
Comment (zimulala) [inline executor/split.go]: Do we need to return an error？

File: executor/split.go [hunk 10]
```diff
@@ -0,0 +1,82 @@
...
+
+	}
+	if !e.ctx.GetSessionVars().WaitTableSplitFinish {
+		return nil
+	}
+	for _, regionID := range regionIDs {
+		err := s.WaitScatterRegionFinish(regionID)
+		if err != nil {
+			logutil.Logger(context.Background()).Warn("wait scatter region failed",
+				zap.Uint64("regionID", regionID),
```
Comment (zimulala) [inline executor/split.go]: Do we need to add the table and index information to this log？

File: go.mod [hunk 11]
```diff
@@ -73,3 +73,5 @@ require (
 	sourcegraph.com/sourcegraph/appdash v0.0.0-20180531100431-4c381bd170b4
 	sourcegraph.com/sourcegraph/appdash-data v0.0.0-20151005221446-73f23eafcf67
 )
+
+replace github.com/pingcap/parser => github.com/crazycs520/parser v0.0.0-20190506044601-7645e4600bc6
```
Comment (zimulala) [inline go.mod]: Please update this.

File: executor/split.go (hunk 9 above)
Comment (crazycs520) [reply executor/split.go]: You mean when where is only 1 tidb-server and no tikv-server, then return error here?

Comment (crazycs520) [reply executor/split.go]: I see https://github.com/pingcap/tidb/blob/release-3.0/ddl/table.go#L346 just return here, so ...

Comment (zimulala) [reply executor/split.go]: Nop. I forgot it, sorry.

File: executor/split.go (hunk 10 above)
Comment (crazycs520) [reply executor/split.go]: done.

**Pull request merged**